### PR TITLE
feat(crons): Add way to persist errors during checkin ingestion

### PR DESCRIPTION
**Pull request by wedamija**
This adds persistence for errors generated during checkin ingestion. We're just going to use Redis for this since this data is ephemeral and we want to avoid the overhead of postgres.

We'll store up to 10 errors per monitor, and for any checkins where we couldn't get the specific monitor, we'll store up to 10 errors per project as well. These sets will only stick around for 1 day, although we could keep them longer if we think it's valuable.

We store a fair bit of data about the checkin - the entire payload, and any details about the error that occurred that we can get. We might want to rework the format of the error payload, but that can happen in a separate pr.

### CODE REVIEW

File: src/sentry/monitors/processing_errors.py [hunk 1]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+    @classmethod
+    def from_dict(cls, data: CheckinProcessingErrorData) -> CheckinProcessingError:
+        return cls(
+            errors=[ProcessingError.from_dict(error) for error in data["errors"]],
+            checkin=CheckinItem.from_dict(data["checkin"]),
+        )
+
+
+class CheckinProcessErrorsManager:
+    def _get_cluster(self):
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:115]: Can we add a type?

File: src/sentry/monitors/processing_errors.py [hunk 2]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+
+    @classmethod
+    def from_dict(cls, data: CheckinProcessingErrorData) -> CheckinProcessingError:
+        return cls(
+            errors=[ProcessingError.from_dict(error) for error in data["errors"]],
+            checkin=CheckinItem.from_dict(data["checkin"]),
+        )
+
+
+class CheckinProcessErrorsManager:
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:114]: Do we actually need the class? Doesn't seem to have any interesting state, just using it as a namespace?

Comment (wedamija) [reply src/sentry/monitors/processing_errors.py:114]: Yeah, mostly seemed useful to keep it as a namespace rather than have functions scattered around. Could be a module instead?

Comment (evanpurkhiser) [reply src/sentry/monitors/processing_errors.py:114]: yeah could be a module. I'm indifferent

File: src/sentry/monitors/processing_errors.py [hunk 3]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+            except Monitor.DoesNotExist:
+                pass
+        if monitor:
+            error_identifier = f"monitor:{monitor.id}"
+        else:
+            error_identifier = f'project:{error.checkin.message["project_id"]}'
+
+        error_key = f"monitors.processing_errors.{error_identifier}"
+        serialized_error = json.dumps(error.to_dict())
+        redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:135]: Isn't this what `_get_cluster` is for?

File: src/sentry/monitors/processing_errors.py [hunk 4]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+        pipeline = redis_client.pipeline(transaction=False)
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:136]: Yagiz would be proud

File: src/sentry/monitors/processing_errors.py [hunk 5]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+            # Attempt to get the monitor from the checkin info if we failed to retrieve it during ingestion
+            try:
+                monitor = Monitor.objects.get(
+                    project_id=error.checkin.message["project_id"],
+                    slug=error.checkin.payload["monitor_slug"],
+                )
+            except Monitor.DoesNotExist:
+                pass
+        if monitor:
+            error_identifier = f"monitor:{monitor.id}"
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:129]:
```suggestion
            error_identifier = self.build_monitor_identifier(monitor)
```

File: src/sentry/monitors/processing_errors.py [hunk 6]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+    def build_monitor_identifier(self, monitor: Monitor) -> str:
+        return f"monitor:{monitor.id}"
+
+    def get_for_monitor(self, monitor: Monitor) -> list[CheckinProcessingError]:
+        return self._get_for_entity(self.build_monitor_identifier(monitor))
+
+    def build_project_identifier(self, project: Project) -> str:
+        return f"project:{project.id}"
+
+    def get_for_project(self, project: Project) -> list[CheckinProcessingError]:
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:152]:
```suggestion
    def get_for_project(self, project: Project) -> Sequence[CheckinProcessingError]:
```

File: src/sentry/monitors/processing_errors.py [hunk 7]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+        return self._get_for_entity(self.build_project_identifier(project))
+
+    def _get_for_entity(self, identifier: str) -> list[CheckinProcessingError]:
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:155]:
```suggestion
    def _get_for_entity(self, identifier: str) -> Sequence[CheckinProcessingError]:
```

File: tests/sentry/monitors/test_types.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+from sentry.monitors.testutils import build_checkin_item
+from sentry.monitors.types import CheckinItem
+from sentry.testutils.cases import TestCase
+
+
+class CheckinItemTest(TestCase):
```
Comment (evanpurkhiser) [inline tests/sentry/monitors/test_types.py:6]: good call. Could be it's own PR if you wanted to pull it out

File: src/sentry/testutils/fixtures.py [hunk 9]
```diff
@@ -460,6 +461,23 @@ def create_notification_settings_provider(self, *args, **kwargs):
     def create_user_option(self, *args, **kwargs):
         return Factories.create_user_option(*args, **kwargs)
 
+    def create_monitor(self, **kwargs):
```
Comment (evanpurkhiser) [inline src/sentry/testutils/fixtures.py:464]: Can probably be reused in a lot of our tests. We have some monitor base test case that has this too

File: src/sentry/monitors/testutils.py [hunk 10]
```diff
@@ -1,7 +1,15 @@
...
+
+from sentry.monitors.processing_errors import (
+    CheckinProcessingError,
+    ProcessingError,
+    ProcessingErrorType,
+)
+from sentry.monitors.types import CheckinItem, CheckinPayload
 
 
 def make_ref_time(**kwargs):
```
Comment (evanpurkhiser) [inline src/sentry/monitors/testutils.py:15]: Unrelated but I need to delete this

Comment (evanpurkhiser) [reply src/sentry/monitors/testutils.py:15]: https://github.com/getsentry/sentry/pull/70429

File: src/sentry/monitors/processing_errors.py [hunk 11]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+        redis = self._get_cluster()
+        error_key = f"monitors.processing_errors.{identifier}"
+        raw_errors = redis.zrange(error_key, 0, MAX_ERRORS_PER_SET, desc=True)
+        return [CheckinProcessingError.from_dict(json.loads(raw_error)) for raw_error in raw_errors]
+
+
+def handle_processing_errors(item: CheckinItem, error: CheckinValidationError):
+    try:
+        organization = Organization.objects.get(project__id=item.message["project_id"])
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:164]: should we get from cache?

File: src/sentry/monitors/processing_errors.py [hunk 12]
```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
...
+                monitor = Monitor.objects.get(
+                    project_id=error.checkin.message["project_id"],
+                    slug=error.checkin.payload["monitor_slug"],
+                )
+            except Monitor.DoesNotExist:
+                pass
+        if monitor:
+            error_identifier = f"monitor:{monitor.id}"
+        else:
+            error_identifier = f'project:{error.checkin.message["project_id"]}'
```
Comment (evanpurkhiser) [inline src/sentry/monitors/processing_errors.py:131]: wish we could use `build_project_identifier`. Tempted to say the signature could be `project_or_project_id` but feels kinda gross

**Review by evanpurkhiser — APPROVED** [review]
Left a few comments but I think it's good to go

File: src/sentry/monitors/processing_errors.py (hunk 3 above)
Comment (wedamija) [reply src/sentry/monitors/processing_errors.py:135]: I added a bunch of convenience methods and forgot to use them 🤦

File: src/sentry/monitors/processing_errors.py (hunk 6 above)
Comment (wedamija) [reply src/sentry/monitors/processing_errors.py:152]: Not that clear on what is preferred here. We know we're returning an actual list, isn't it better to be specific? I see both options in the codebase

File: src/sentry/testutils/fixtures.py (hunk 9 above)
Comment (wedamija) [reply src/sentry/testutils/fixtures.py:464]: Yeah, ideally we'd just use this one method

File: src/sentry/monitors/processing_errors.py (hunk 11 above)
Comment (wedamija) [reply src/sentry/monitors/processing_errors.py:164]: There's no cache for doing this by project id https://github.com/getsentry/sentry/blob/fa0930966ce70d49060dda79c9f951b334d35929/src/sentry/models/organization.py#L203

I could fetch the project by cache, then fetch the org by cache... maybe that's better

File: src/sentry/monitors/processing_errors.py (hunk 6 above)
Comment (evanpurkhiser) [reply src/sentry/monitors/processing_errors.py:152]: You're right https://typing.readthedocs.io/en/latest/source/best_practices.html#arguments-and-return-types

> For return values, prefer concrete types (list, dict, etc.) for concrete implementations. The return values of protocols and abstract base classes must be judged on a case-by-case basis.
